doc/vmips.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "VMIPS 1"
.TH VMIPS 1 "2007-03-27" "vmips 1.3.2" "VMIPS Programmer's Manual"
.SH "NAME"
vmips \- VMIPS R3000\-based MIPS simulator (running and customizing)
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 4
\&        vmips [-n] [-F FILE] [-o option_string] ... rom_file
\&        vmips --help
\&        vmips --version
\&        vmips --print-config
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
When you invoke vmips, it loads a \s-1ROM\s0 file specified on the command line,
initializes and resets the \s-1MIPS\s0 processor, and begins execution at the first
address of the \s-1ROM\s0 file, which is normally loaded at address 0xbfc00000.
.SH "OPTIONS"
.IX Header "OPTIONS"
This is what the different command line options mean:
.IP "\fB\-F \s-1FILE\s0\fR" 4
.IX Item "-F FILE"
Read options from \s-1FILE\s0 instead of the \fI.vmipsrc\fR in your home directory.
.IP "\fB\-n\fR" 4
.IX Item "-n"
Do not read the system-wide configuration file, usually called
\&\fI/usr/local/etc/vmipsrc\fR.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
Prints a short summary of \s-1VMIPS\s0 command line options, and exits
successfully.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Prints a short summary of \s-1VMIPS\s0 version and copyright information,
and exits successfully.
.IP "\fB\-\-print\-config\fR" 4
.IX Item "--print-config"
Prints a short summary of \s-1VMIPS\s0 compile-time configuration information,
and exits successfully.
.IP "\fB\-o something\fR" 4
.IX Item "-o something"
Set the option \*(L"something\*(R" as if \*(L"something\*(R" were in your
\&.vmipsrc file.  See the \*(L"\s-1VMIPS\s0 options\*(R" section of the
\&\*(L"Customizing\*(R" chapter for more information on what kind
of things can go in your .vmipsrc file. You can use as many
\&\-o options on the command line as your shell will let you.
.IP "\fBrom_file\fR" 4
.IX Item "rom_file"
Use the named file as the \s-1ROM\s0 file \s-1VMIPS\s0 should boot. This
option is mandatory.
.SH "FILES"
.IX Header "FILES"
.Sh "\s-1VMIPS\s0 options"
.IX Subsection "VMIPS options"
The \s-1VMIPS\s0 simulator gets runtime options from four different sources,
in this order: first, it checks its compile-time defaults, which are
set by the site administrator in the source file \fIoptiontbl.h\fR. Then,
the system-wide configuration file is read, unless you specify the \fB\-n\fR
option; usually this file is called
\&\fI/usr/local/etc/vmipsrc\fR, but it may have been moved by the site
administrator or by the maintainers of your distribution.
(This is configurable in the source file \fIoptions.h\fR, and
by specifying the \-\-prefix and \-\-sysconfdir options to the \s-1GNU\s0
\&\fBconfigure\fR
script when building \s-1VMIPS\s0.) Next, it checks the user's own configuration
file, usually the file \fI.vmipsrc\fR in your home directory, or whatever
file you specify using the \fB\-F\fR option. Last, it
reads the command line, and gets any options listed there.
.Sh "Format of the configuration file"
.IX Subsection "Format of the configuration file"
The configuration file may contain as many options per line as you want,
provided no single line exceeds 1,024 characters in length.  Whitespace
separates options from one another.  Single quotes and backslash are valid
in the configuration file. Their meanings are similar to those found in
the Bourne shell: any text within paired single quotes is uninterpreted,
as is any character immediately following a backslash.  A comment is
any text starting from a hash mark to the end of the line, inclusive.
.PP
A string or number option named \s-1NAME\s0 can appear as NAME=VALUE, where \s-1VALUE\s0
is the string or number in question.  If the number begins with 0x, it will
be interpreted as a 32\-bit hexadecimal number, and if it begins with 0,
it will be interpreted as octal. Otherwise, it will be interpreted as a
decimal number. Numbers are always unsigned. A Boolean option named \s-1NAME\s0
can appear as either \s-1NAME\s0 (to set it to \s-1TRUE\s0) or noNAME (to set it to \s-1FALSE\s0).
.Sh "Summary of configuration options"
.IX Subsection "Summary of configuration options"
The following is a list of the configuration options present in this
version of \s-1VMIPS\s0.
.PP
\&\fBhaltdumpcpu\fR (type: Boolean)
.PP
.Vb 3
\& Controls whether the CPU registers and stack will be dumped
\&on halt. For the output format, please see the description of the
\&B<dumpcpu> option, below.  The default value is FALSE.
.Ve
.PP
\&\fBhaltdumpcp0\fR (type: Boolean)
.PP
.Vb 4
\& Controls whether the system control coprocessor (CP0) registers
\&and the contents of the translation lookaside buffer (TLB) will be
\&dumped on halt. For the output format, please see the description
\&of the B<dumpcp0> option, below.  The default value is FALSE.
.Ve
.PP
\&\fBexcpriomsg\fR (type: Boolean)
.PP
.Vb 4
\& Controls whether exception prioritizing messages will
\&be printed.  These messages attempt to explain which of
\&a number of exceptions caused by the same instruction
\&will be reported.  The default value is FALSE.
.Ve
.PP
\&\fBexcmsg\fR (type: Boolean)
.PP
.Vb 8
\& Controls whether every exception will cause a message
\&to be printed. The message gives the exception code, a
\&short explanation of the exception code, its priority,
\&the delay slot state of the virtual CPU, and states
\&what type of memory access the exception was caused by,
\&if applicable. Interrupt exceptions are only printed if
\&B<reportirq> is also set; when they occur, they also have Cause
\&and Status register information printed.  The default value is FALSE.
.Ve
.PP
\&\fBbootmsg\fR (type: Boolean)
.PP
.Vb 3
\& Controls whether boot-time and halt-time messages will be printed.
\&These include ROM image size, self test messages, reset and halt
\&announcements, and possibly other messages.  The default value is TRUE.
.Ve
.PP
\&\fBinstdump\fR (type: Boolean)
.PP
.Vb 2
\& Controls whether every instruction executed will be disassembled
\&and printed.  The default value is FALSE. The output is in the following format:
.Ve
.PP
.Vb 1
\&        PC=0xbfc00000 [1fc00000]    24000000 li $zero,0
.Ve
.PP
The first column contains the \s-1PC\s0 (program counter), followed by
the physical translation of that address in brackets. The third
column contains the machine instruction word at that address,
followed by the assembly language corresponding to that word.
All of the constants except for the assembly language are in
hexadecimal. 
.PP
\&\fBdumpcpu\fR (type: Boolean)
.PP
.Vb 2
\& Controls whether the CPU registers and stack will be dumped after every
\&instruction.  The default value is FALSE. The output is in the following format:
.Ve
.PP
.Vb 6
\&        Reg Dump: [ PC=bfc00180  LastInstr=0000000d  HI=00000000  LO=00000000
\&                    DelayState=NORMAL  DelayPC=bfc00308  NextEPC=bfc00308
\&         R00=00000000  R01=00000000  R02=00000000  R03=a00c000e  R04=0000000a 
\&         ...
\&         R30=00000000  R31=bfc00308  ]
\&        Stack: 00000000 00000000 00000000 00000000 a2000008 a2000008 ...
.Ve
.PP
(Some values have been omitted for brevity.)
Here, \s-1PC\s0 is the program counter, LastInstr is the last instruction
executed, \s-1HI\s0 and \s-1LO\s0 are the multiplication/division result registers,

NEWS

Appendix D Release History
**************************

   vmips-1.3.2 was released on 20 July 2005.  User-visible changes in
version 1.3.2 (since version 1.3.1):

   * A bug was fixed in the code which responded to address-translation
     exceptions by loading the CP0 Context register. It was loading the
     wrong virtual page number.

   * Problems compiling VMIPS with gcc 4.1.x have been fixed.

   vmips-1.3.1 was released on 5 January 2005.  User-visible changes in
version 1.3.1 (since version 1.3):

   * VMIPS now compiles and passes most of its regression tests under
     Microsoft Windows using Cygwin. There is one known failure
     (misc-tests/echo), but the failure stems from the testsuite, not
     VMIPS; I suspect that my old version of Expect is at fault.

   * The build status report script, `buildstat.sh', has been removed
     from the distribution.  (Since it was broken in version 1.3 and no
     one complained, we're assuming that no one will miss it much!)

   * Two bugs in the testsuite that could have led to spurious test
     failures have been fixed.

   vmips-1.3 was released on 8 October 2004.  User-visible changes in
version 1.3 (since version 1.2.2):

   * The VMIPS code base has been cleaned up substantially. The most
     important visible effect of this is that VMIPS compiles much more
     quickly now.

   * A new, experimental executable-file loader has been integrated.
     Presently it is only usable for loading ECOFF files into the
     kernel segments (0x80000000-0xbfffffff). The ROM monitor has a new
     `call' command you can use to take advantage of this.

   * Vmipstool can now disassemble binary instructions from the command
     line, e.g.  `vmipstool --disassemble-word <pc> <instr>'.

   * The `excmsg' option used to report all exceptions, including
     things like clock interrupts. This made using `excmsg' in any
     nontrivial kernel code almost impossible. Now you must also specify
     'reportirq' to get reports of interrupt exceptions.

   * The `haltjrra' option has been removed.

   * The DECstation 5000/200-compatible CSR device now supports delivery
     of interrupts from devices attached to CSR interrupt lines, such
     as the DZ11 serial device.

   * The DZ11 serial device now supports interrupt-driven I/O properly.
     It previously only worked correctly if polling was used.

   * The documentation has received the usual slight adjustments. In
     addition, the test-suite documentation has been extensively
     revised and integrated into the programmer's manual.

   * The options-processing code has been revised slightly to print a
     better error message when an invalid option is specified.

   vmips-1.2.2 was released on 23 August 2004.  User-visible changes in
version 1.2.2 (since version 1.2.1):

   * A bug in the Makefiles and configure scripts has been fixed, where
     some files would not be installed if you didn't configure with
     cross compilation tools (using `--with-mips').

   * Vladimir Machulsky found and fixed a bug in the debugger interface,
     where the register file would be sent to gdb incorrectly. Thanks!

   * VMIPS will no longer abort and dump core when bad command-line
     rguments are passed to it.  Instead, it will print an error
     message and exit with a non-zero exit code.

   * A bug in the command-line option processor was fixed, where using a
     completely empty `~/.vmipsrc' or the `-F /dev/null' option could
     cause the options parser to parse garbage, usually resulting in
     harmless, but annoying error messages.

   vmips-1.2.1 was released on 26 July 2004.  User-visible changes in
version 1.2.1 (since version 1.2):

   * Fixed a bug where delay slot execution could be triggered even when
     emulating a malformed jr (jump-register) instruction.

   * Fixed a bug where, upon taking a TLB Miss exception, the ASID of
     the page that could not be looked up would be overwritten (in the
     CP0 EntryHi register) with garbage.

   * Fixed a bug in VMIPS configurations with multiple interrupt
     devices, where Interrupt Pending bits of the CP0 Cause from
     previous interrupts could be left asserted even if the device that
     asserted them had subsequently deasserted them.

   * Fixed a bug where the memory dump file and ROM file would be opened
     in text mode. They are now opened in binary mode.

   vmips-1.2 was released on 26 June 2004.  User-visible changes in
version 1.2 (since version 1.1.3):

   * Many documentation updates have been made. Note that the
     documentation has been relicensed under the MIT license, instead
     of the GNU FDL.

   * Many facets of VMIPS are now configurable at runtime, instead of at
     compile time. For instance:

     	It is now possible to build and install VMIPS without having
     previously installed MIPS cross-compiler tools. In particular,
     VMIPS now incorporates the portions of GNU libopcodes used to
     implement the MIPS disassembler, so linking against an installed
     version of libopcodes from GNU binutils is no longer necessary.

     	It is now possible to switch the VMIPS CPU from big-endian to
     little-endian mode or vice-versa using a command-line option.

     	It is now possible to change the name of the configuration file
     that VMIPS reads on startup using a command-line option. The "-o
     configfile" option, which never worked, has been removed.


vmips.cc
/* Main driver program for VMIPS.
   Copyright 2001, 2003 Brian R. Gaeke.
   Copyright 2002, 2003 Paul Twohey.

This file is part of VMIPS.

VMIPS is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

VMIPS is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with VMIPS; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#include "clock.h"
#include "clockdev.h"
#include "clockreg.h"
#include "cpzeroreg.h"
#include "debug.h"
#include "error.h"
#include "endiantest.h"
#include "haltreg.h"
#include "haltdev.h"
#include "intctrl.h"
#include "range.h"
#include "spimconsole.h"
#include "mapper.h"
#include "memorymodule.h"
#include "cpu.h"
#include "cpzero.h"
#include "spimconsreg.h"
#include "vmips.h"
#include "options.h"
#include "decrtc.h"
#include "decrtcreg.h"
#include "deccsr.h"
#include "deccsrreg.h"
#include "decstat.h"
#include "decserial.h"
#include "stub-dis.h"
#include "rommodule.h"
#include <fcntl.h>
#include <cerrno>
#include <cassert>
#include <csignal>
#include <cstdarg>
#include <string>
#include <exception>
#include <string.h>

vmips *machine;

void
vmips::refresh_options(void)
{
	/* Extract important flags and things. */
	opt_bootmsg = opt->option("bootmsg")->flag;
	opt_clockdevice = opt->option("clockdevice")->flag;
	opt_debug = opt->option("debug")->flag;
	opt_dumpcpu = opt->option("dumpcpu")->flag;
	opt_dumpcp0 = opt->option("dumpcp0")->flag;
	opt_haltdevice = opt->option("haltdevice")->flag;
	opt_haltdumpcpu = opt->option("haltdumpcpu")->flag;
	opt_haltdumpcp0 = opt->option("haltdumpcp0")->flag;
	opt_instcounts = opt->option("instcounts")->flag;
	opt_memdump = opt->option("memdump")->flag;
	opt_realtime = opt->option("realtime")->flag;
 
	opt_clockspeed = opt->option("clockspeed")->num;
	clock_nanos = 1000000000/opt_clockspeed;

	opt_clockintr = opt->option("clockintr")->num;
	opt_clockdeviceirq = opt->option("clockdeviceirq")->num;
	opt_loadaddr = opt->option("loadaddr")->num;
	opt_memsize = opt->option("memsize")->num;
	opt_timeratio = opt->option("timeratio")->num;
 
	opt_memdumpfile = opt->option("memdumpfile")->str;
	opt_image = opt->option("romfile")->str;
	opt_execname = opt->option("execname")->str;
	opt_ttydev = opt->option("ttydev")->str;
	opt_ttydev2 = opt->option("ttydev2")->str;
	opt_spimconsole = opt->option("spimconsole")->flag;

	opt_decrtc = opt->option("decrtc")->flag;
	opt_deccsr = opt->option("deccsr")->flag;
	opt_decstat = opt->option("decstat")->flag;
	opt_decserial = opt->option("decserial")->flag;
}

/* Set up some machine globals, and process command line arguments,
 * configuration files, etc.
 */
vmips::vmips(int argc, char *argv[])
	: opt(new Options), halted(false),
	  clock(0), clock_device(0), halt_device(0), spim_console(0),
	  num_instrs(0)
{
   	opt->process_options (argc, argv);
	refresh_options();
}

vmips::~vmips() throw()
{
	delete spim_console;
	delete halt_device;
	delete clock_device;
	delete clock;
}

void
vmips::setup_machine(void)
{
	/* Construct the various vmips components. */
	intc = new IntCtrl;
	physmem = new Mapper;
	cpu = new CPU (*physmem, *intc);

	/* Set up the debugger interface, if applicable. */
	if (opt_debug)
		dbgr = new Debug (*cpu, *physmem);

    /* Direct the libopcodes disassembler output to stderr. */
    disasm = new Disassembler (host_bigendian, stderr);
}

/* Connect the file or device named NAME to line number L of
 * console device C, or do nothing if NAME is "off".
 */
void vmips::setup_console_line(int l, char *name, TerminalController *c, const
char *c_name) throw()
{
	/* If they said to turn off the tty line, do nothing. */
	if (strcmp(name, "off") == 0)
		return;

	/* Open the file or device in question. */
	int ttyfd = open(name, O_RDWR | O_NONBLOCK);
	if (ttyfd == -1) {
		/* If we can't open it, warn and use stdout instead. */
		error("Opening %s (terminal %d): %s", name, l, strerror(errno));
		warning("using stdout, input disabled\n");
		ttyfd = fileno(stdout);
	}

	/* Connect it to the SPIM-compatible console device. */
	c->connect_terminal(ttyfd, l);
	boot_msg("Connected fd %d to %s line %d.\n", ttyfd, c_name, l);
}

bool vmips::setup_spimconsole() throw( std::bad_alloc )
{
	/* FIXME: It would be helpful to restore tty modes on a SIGINT or
	   other abortive exit or when vmips has been foregrounded after
	   being in the background. The restoration mechanism should use
	   TerminalController::reinitialze_terminals() */

	if (!opt_spimconsole)
		return true;
	
	spim_console = new SpimConsoleDevice( clock );
	physmem->map_at_physical_address( spim_console, SPIM_BASE );
	boot_msg("Mapping %s to physical address 0x%08x\n",
		  spim_console->descriptor_str(), SPIM_BASE);
	
	intc->connectLine(IRQ2, spim_console);
	intc->connectLine(IRQ3, spim_console);
	intc->connectLine(IRQ4, spim_console);
	intc->connectLine(IRQ5, spim_console);
	intc->connectLine(IRQ6, spim_console);
	boot_msg("Connected IRQ2-IRQ6 to %s\n",spim_console->descriptor_str());

	setup_console_line(0, opt_ttydev, spim_console,
      spim_console->descriptor_str ());
	setup_console_line(1, opt_ttydev2, spim_console,
      spim_console->descriptor_str ());
	return true;
}

bool vmips::setup_clockdevice() throw( std::bad_alloc )
{
	if( !opt_clockdevice )
		return true;

	uint32 clock_irq;
	if( !(clock_irq = DeviceInt::num2irq( opt_clockdeviceirq )) ) {
		error( "invalid clockdeviceirq (%u), irq numbers must be 2-7.",
		       opt_clockdeviceirq );
		return false;
	}	

	/* Microsecond Clock at base physaddr CLOCK_BASE */
	clock_device = new ClockDevice( clock, clock_irq, opt_clockintr );
	physmem->map_at_physical_address( clock_device, CLOCK_BASE );
	boot_msg( "Mapping %s to physical address 0x%08x\n",
		  clock_device->descriptor_str(), CLOCK_BASE );

	intc->connectLine( clock_irq, clock_device );
	boot_msg( "Connected %s to the %s\n", DeviceInt::strlineno(clock_irq),
		  clock_device->descriptor_str() );

	return true;
}

bool vmips::setup_decrtc() throw( std::bad_alloc )
{
	if (!opt_decrtc)
		return true;

	/* Always use IRQ3 ("hw interrupt level 1") for RTC. */
	uint32 decrtc_irq = DeviceInt::num2irq (3);

	/* DECstation 5000/200 DS1287-based RTC at base physaddr DECRTC_BASE */
	decrtc_device = new DECRTCDevice( clock, decrtc_irq );
	physmem->map_at_physical_address( decrtc_device, DECRTC_BASE );
	boot_msg( "Mapping %s to physical address 0x%08x\n",
		  decrtc_device->descriptor_str(), DECRTC_BASE );

	intc->connectLine( decrtc_irq, decrtc_device );
	boot_msg( "Connected %s to the %s\n", DeviceInt::strlineno(decrtc_irq),
		  decrtc_device->descriptor_str() );

	return true;
}

bool vmips::setup_deccsr() throw( std::bad_alloc )
{
	if (!opt_deccsr)
		return true;

	/* DECstation 5000/200 Control/Status Reg at base physaddr DECCSR_BASE */
    /* Connected to IRQ2 */
    static const uint32 DECCSR_MIPS_IRQ = DeviceInt::num2irq (2);
	deccsr_device = new DECCSRDevice (DECCSR_MIPS_IRQ);
	physmem->map_at_physical_address (deccsr_device, DECCSR_BASE);
	boot_msg ("Mapping %s to physical address 0x%08x\n",
		  deccsr_device->descriptor_str(), DECCSR_BASE);

	intc->connectLine (DECCSR_MIPS_IRQ, deccsr_device);
    boot_msg("Connected %s to the %s\n", DeviceInt::strlineno(DECCSR_MIPS_IRQ),
             deccsr_device->descriptor_str());

	return true;
}

bool vmips::setup_decstat() throw( std::bad_alloc )
{
	if (!opt_decstat)
		return true;

	/* DECstation 5000/200 CHKSYN + ERRADR at base physaddr DECSTAT_BASE */
	decstat_device = new DECStatDevice( );
	physmem->map_at_physical_address( decstat_device, DECSTAT_BASE );
	boot_msg( "Mapping %s to physical address 0x%08x\n",
		  decstat_device->descriptor_str(), DECSTAT_BASE );

	return true;
}

bool vmips::setup_decserial() throw( std::bad_alloc )
{
	if (!opt_decserial)
		return true;

	/* DECstation 5000/200 DZ11 serial at base physaddr DECSERIAL_BASE */
	/* Uses CSR interrupt SystemInterfaceCSRInt */
	decserial_device = new DECSerialDevice (clock, SystemInterfaceCSRInt);
	physmem->map_at_physical_address (decserial_device, DECSERIAL_BASE );
	boot_msg ("Mapping %s to physical address 0x%08x\n",
		  decserial_device->descriptor_str (), DECSERIAL_BASE );

	// Use printer line for console.
	setup_console_line (3, opt_ttydev, decserial_device,
      decserial_device->descriptor_str ());
	return true;
}

bool vmips::setup_haltdevice() throw( std::bad_alloc )
{
	if( !opt_haltdevice )
		return true;

	halt_device = new HaltDevice( this );
	physmem->map_at_physical_address( halt_device, HALT_BASE );
	boot_msg( "Mapping %s to physical address 0x%08x\n",
		  halt_device->descriptor_str(), HALT_BASE );

	return true;
}

void vmips::boot_msg( const char *msg, ... ) throw()
{
	if( !opt_bootmsg )
		return;

	va_list ap;
	va_start( ap, msg );
	vfprintf( stderr, msg, ap );
	va_end( ap );

	fflush( stderr );
}

int
vmips::host_endian_selftest(void)
{
  try {
    EndianSelfTester est;
    machine->host_bigendian = est.host_is_big_endian();
    if (!machine->host_bigendian) {
      boot_msg ("Little-Endian host processor detected.\n");
    } else {
      boot_msg ("Big-Endian host processor detected.\n");
    }
    return 0;
  } catch (std::string &err) {
    boot_msg (err.c_str ());
    return -1;
  }
}

void
vmips::halt(void) throw()
{
	halted = true;
}

uint32
vmips::get_file_size (FILE *fp) throw ()
{
	off_t here, there;

	assert (fp && "Null pointer passed to vmips::get_file_size ()");
	here = ftell (fp);
	fseek (fp, 0, SEEK_END);
	there = ftell (fp);
	fseek (fp, here, SEEK_SET);
	return there - here;
}

void
vmips::step(void)
{
	/* Process instructions. */
	cpu->step();

	/* Keep track of time passing. Each instruction either takes
	 * clock_nanos nanoseconds, or we use pass_realtime() to check the
	 * system clock.
     */
	if( !opt_realtime )
	   clock->increment_time(clock_nanos);
	else
	   clock->pass_realtime(opt_timeratio);

	/* If user requested it, dump registers from CPU and/or CP0. */
	if (opt_dumpcpu)
		cpu->dump_regs_and_stack(stderr);
	if (opt_dumpcp0)
		cpu->cpzero_dump_regs_and_tlb(stderr);
	num_instrs++;
}

long 
timediff(struct timeval *after, struct timeval *before)
{
    return (after->tv_sec * 1000000 + after->tv_usec) -
        (before->tv_sec * 1000000 + before->tv_usec);
}

bool
vmips::setup_rom ()
{
  // Open ROM image.
  FILE *rom = fopen (opt_image, "rb");
  if (!rom) {
    error ("Could not open ROM `%s': %s", opt_image, strerror (errno));
    return false;
  }
  // Translate loadaddr to physical address.
  opt_loadaddr -= KSEG1_CONST_TRANSLATION;
  ROMModule *rm;
  try {
    rm = new ROMModule (rom);
  } catch (int errcode) {
    error ("mmap failed for %s: %s", opt_image, strerror (errcode));
    return false;
  }
  // Map the ROM image to the virtual physical memory.
  physmem->map_at_physical_address (rm, opt_loadaddr);
  boot_msg ("Mapping ROM image (%s, %u words) to physical address 0x%08x\n",
            opt_image, rm->getExtent () / 4, rm->getBase ());
  // Point debugger at wherever the user thinks the ROM is.
  if (opt_debug)
    if (dbgr->setup (opt_loadaddr, rm->getExtent () / 4) < 0)
      return false; // Error in setting up debugger.
  return true;
}

bool
vmips::setup_ram () throw( std::bad_alloc )
{
  // Make a new RAM module and install it at base physical address 0.
  memmod = new MemoryModule(opt_memsize);
  physmem->map_at_physical_address(memmod, 0);
  boot_msg( "Mapping RAM module (host=%p, %uKB) to physical address 0x%x\n",
	    memmod->getAddress (), memmod->getExtent () / 1024, memmod->getBase ());

//init heap
//	....
  return true;
}

bool
vmips::setup_clock () throw( std::bad_alloc )
{
  /* Set up the clock with the current time. */
  timeval start;
  gettimeofday(&start, NULL);
  timespec start_ts;
  TIMEVAL_TO_TIMESPEC( &start, &start_ts );
  clock = new Clock( start_ts );
  return true;
}

static void halt_machine_by_signal(int i)
{
	machine->halt();
}

int
vmips::run()
{
	/* Check host processor endianness. */
	if (host_endian_selftest () != 0) {
		error( "Could not determine host processor endianness." );
		return 1;
	}

	/* Set up the rest of the machine components. */
	setup_machine();

	if (!setup_rom ()) 
	  return 1;

	if (!setup_ram ())
	  return 1;

	if (!setup_haltdevice ())
	  return 1;

	if (!setup_clock ())
	  return 1;

	if (!setup_clockdevice ())
	  return 1;

	if (!setup_decrtc ())
	  return 1;

	if (!setup_deccsr ())
	  return 1;

	if (!setup_decstat ())
	  return 1;

	if (!setup_decserial ())
	  return 1;

	if (!setup_spimconsole ())
	  return 1;

    signal (SIGQUIT, halt_machine_by_signal);

	/* Reset the CPU. */
	boot_msg( "\n*************RESET*************\n\n" );
	cpu->reset();

	if (!setup_exe ())
	  return 1;

	timeval start;
	if (opt_instcounts)
		gettimeofday(&start, NULL);

	if (opt_debug) {
		dbgr->serverloop();
	} else {
		while (! halted)
			step();
	}

	timeval end;
	if (opt_instcounts)
		gettimeofday(&end, NULL);

	/* Halt! */
	boot_msg( "\n*************HALT*************\n\n" );

    /* If we're tracing, dump the trace. */
    cpu->maybe_dump_trace ();

	/* If user requested it, dump registers from CPU and/or CP0. */
	if (opt_haltdumpcpu || opt_haltdumpcp0) {
		fprintf(stderr,"Dumping:\n");
		if (opt_haltdumpcpu)
			cpu->dump_regs_and_stack(stderr);
		if (opt_haltdumpcp0)
			cpu->cpzero_dump_regs_and_tlb(stderr);
	}

	if (opt_instcounts) {
		double elapsed = (double) timediff(&end, &start) / 1000000.0;
		fprintf(stderr, "%u instructions in %.5f seconds (%.3f "
			"instructions per second)\n", num_instrs, elapsed,
			((double) num_instrs) / elapsed);
	}

	if (opt_memdump) {
		fprintf(stderr,"Dumping RAM to %s...", opt_memdumpfile);
		if (FILE *ramdump = fopen (opt_memdumpfile, "wb")) {
			fwrite (memmod->getAddress (), memmod->getExtent (), 1, ramdump);
			fclose(ramdump);
			fprintf(stderr,"succeeded.\n");
		} else {
			error( "\nRAM dump failed: %s", strerror(errno) );
		}
	}

	/* We're done. */
	boot_msg( "Goodbye.\n" );
	return 0;
}

static void vmips_unexpected() {
  fatal_error ("unexpected exception");
}

static void vmips_terminate() {
  fatal_error ("uncaught exception");
}
/*
int
main(int argc, char **argv)
{
	try {
		std::set_unexpected(vmips_unexpected);
		std::set_terminate(vmips_terminate);

		machine = new vmips(argc, argv);
		int rc = machine->run();
		delete machine; /* No disassemble Number Five!! */
/*		return rc;
	}
	catch( std::bad_alloc &b ) {
		fatal_error( "unable to allocate memory" );
	}
}
*/